modules/service/src/main/scala/lucuma/odb/service/ItcClientService.scala

```diff
@@ -0,0 +1,136 @@
+
+package lucuma.odb.service
+
+import cats.data.NonEmptyList
+import cats.effect.Sync
+import cats.syntax.applicative.*
+import cats.syntax.flatMap.*
+import cats.syntax.functor.*
+import edu.gemini.grackle.Result
+import lucuma.core.enums.Band
+import lucuma.core.math.BrightnessUnits.BrightnessMeasure
+import lucuma.core.math.BrightnessUnits.Integrated
+import lucuma.core.math.BrightnessUnits.Surface
+import lucuma.core.math.Wavelength
+import lucuma.core.model.Observation
+import lucuma.core.model.Program
+import lucuma.core.model.SourceProfile
+import lucuma.core.model.Target
+import lucuma.core.model.User
+import lucuma.itc.client.InstrumentMode
+import lucuma.itc.client.SpectroscopyModeInput
+import skunk.Session
+
+import scala.collection.immutable.SortedMap
+
+
+/*
+final case class SpectroscopyModeInput(
+  wavelength:     Wavelength,
+  signalToNoise:  PosBigDecimal,
+  sourceProfile:  SourceProfile,
+  band:           Band,
+  radialVelocity: RadialVelocity,
+  constraints:    ConstraintSet,
+  mode:           InstrumentMode
+)
+*/
+
+trait ItcClientService[F[_]] {
+
+  def selectSpectroscopyInput(
+    programId: Program.Id,
+    which:     List[Observation.Id]
+  ): F[Map[Observation.Id, NonEmptyList[SpectroscopyModeInput]]]
+
+}
+
+object ItcClientService {
+
+  def fromSession[F[_]: Sync](
+    session:  Session[F],
+    user:     User,
+    oService: ObservationService[F],
+    mService: ObservingModeServices[F],
+    aService: AsterismService[F],
+    tService: TargetService[F]
+  ): ItcClientService[F] =
+
+    new ItcClientService[F] {
+
+      private def spectroscopy(
+        o: ObservationService.ItcParams,
+        m: ObservingModeServices.ItcParams,
+        t: TargetService.ItcParams
+      ): Option[SpectroscopyModeInput] = {
+
+        def extractBand[T](w: Wavelength, bMap: SortedMap[Band, BrightnessMeasure[T]]): Option[Band] =
+          bMap.minByOption { case (b, _) =>
+            (w.toPicometers.value.value - b.center.toPicometers.value.value).abs
+          }.map(_._1)
+
+        def band: Option[Band] =
+          SourceProfile
+            .integratedBrightnesses
+            .getOption(t.sourceProfile)
+            .flatMap(bMap => extractBand[Integrated](m.wavelength, bMap))
+            .orElse(
+              SourceProfile
+                .surfaceBrightnesses
+                .getOption(t.sourceProfile)
+                .flatMap(bMap => extractBand[Surface](m.wavelength, bMap))
+            )
+
+        band.map { b =>
+          SpectroscopyModeInput(
+            m.wavelength,
+            o.signalToNoise,
+            o.signalToNoiseAt,
+            t.sourceProfile,
+            b,
+            t.radialVelocity,
+            o.constraints,
+            m.mode
+          )
+        }
+      }
+
+      override def selectSpectroscopyInput(
+        programId: Program.Id,
+        which:     List[Observation.Id]
+      ): F[Map[Observation.Id, NonEmptyList[SpectroscopyModeInput]]] =
```

Comment:
This pulls all the information together.  It's rather involved.  At the moment observations for which there is insufficient information for doing the query are just left out of the return map.  I'm not sure whether to report errors or what.